core/hotspot/cache/wtinylfu/tinylfu.go

```diff
@@ -0,0 +1,223 @@
+	doorkeeperFactor         = 8
+	countersFactor           = 1
+	falsePositiveProbability = 0.1
+	lruRatio                 = 0.01
+)
+
+// TinyLfu is an implementation of the TinyLfu algorithm: https://arxiv.org/pdf/1512.00727.pdf
+//           Window Cache Victim .---------. Main Cache Victim
+//          .------------------->| TinyLFU |<-----------------.
+//          |                    `---------'                  |
+// .-------------------.              |    .------------------.
+// | Window Cache (1%) |              |    | Main Cache (99%) |
+// |      (LRU)        |              |    |      (SLRU)      |
+// `-------------------'              |    `------------------'
+//          ^                         |               ^
+//          |                         `---------------'
+//       new item                        Winner
+type TinyLfu struct {
+	countMinSketch *countMinSketch
+	doorkeeper     *doorkeeper
+	additions      int
+	samples        int
+	lru            *lru
+	slru           *slru
+	items          map[interface{}]*list.Element
+	stats          *stats.CacheStats
+}
+
+func NewTinyLfu(cap int, isRecordingStats bool) (*TinyLfu, error) {
+	if cap <= 0 {
+		return nil, errors.New("Must provide a positive size")
+	}
+	if cap < 100 {
+		cap = 100
+	}
+	lruCap := int(float64(cap) * lruRatio)
+	slruSize := cap - lruCap
+	var statsCache *stats.CacheStats
+	if isRecordingStats {
+		statsCache = stats.NewCacheStats()
+	}
+	items := make(map[interface{}]*list.Element)
+	return &TinyLfu{
+		countMinSketch: newCountMinSketch(countersFactor * cap),
+		additions:      0,
+		samples:        samplesFactor * cap,
+		doorkeeper:     newDoorkeeper(doorkeeperFactor*cap, falsePositiveProbability),
+		items:          items,
+		lru:            newLRU(lruCap, items),
+		slru:           newSLRU(slruSize, items),
+		stats:          statsCache,
+	}, nil
+}
+
+// Get looks up a key's value from the cache.
+func (t *TinyLfu) Get(key interface{}) (interface{}, bool) {
+	return t.get(key, false)
+}
+
+func (t *TinyLfu) get(key interface{}, isInternal bool) (interface{}, bool) {
+	t.additions++
+	if t.additions == t.samples {
+		t.countMinSketch.reset()
+		t.doorkeeper.reset()
+		t.additions = 0
+	}
+
+	val, ok := t.items[key]
+	if !ok {
+		keyHash := sum(key)
+		if t.doorkeeper.put(keyHash) {
+			t.countMinSketch.add(keyHash)
+		}
+		if !isInternal && t.stats != nil {
+			t.stats.RecordMisses()
+		}
+		return nil, false
+	}
+	item := val.Value.(*slruItem)
+	if t.doorkeeper.put(item.keyHash) {
+		t.countMinSketch.add(item.keyHash)
+	}
+
+	v := item.value
+	if item.listType == admissionWindow {
+		t.lru.access(val)
+	} else {
+		t.slru.access(val)
+	}
+	if !isInternal && t.stats != nil {
+		t.stats.RecordHits()
+	}
+	return v, true
+}
+
+// Contains checks if a key is in the cache without updating
+func (t *TinyLfu) Contains(key interface{}) (ok bool) {
+	_, ok = t.items[key]
+	return ok
+}
+
+func (t *TinyLfu) Add(key interface{}, val interface{}) {
```

Comment:
Maybe the implementation `Add` make no sense?

Reply:
This is to be consistent with lru, and there is also this method in the ConcurrentCounterCache interface